drivers/counter/counter_mcux_qtmr.c

```diff
@@ -0,0 +1,332 @@
+
+	return 0;
+}
+
+static int mcux_qtmr_get_value(const struct device *dev, uint32_t *ticks)
+{
+	const struct mcux_qtmr_config *config = dev->config;
+
+	*ticks = QTMR_GetCurrentTimerCount(config->base, config->channel);
+	return 0;
+}
+
+static int mcux_qtmr_set_alarm(const struct device *dev, uint8_t chan_id,
+			      const struct counter_alarm_cfg *alarm_cfg)
+{
+	const struct mcux_qtmr_config *config = dev->config;
+	struct mcux_qtmr_data *data = dev->data;
+	uint32_t current;
+	uint32_t ticks;
+
+	if (chan_id != 0) {
+		LOG_ERR("Invalid channel id");
+		return -EINVAL;
+	}
+
+	if (data->alarm_callback) {
+		return -EBUSY;
+	}
+
+	data->alarm_callback = alarm_cfg->callback;
+	data->alarm_user_data = alarm_cfg->user_data;
+
+	current = QTMR_GetCurrentTimerCount(config->base, config->channel);
+	ticks = alarm_cfg->ticks;
+
+	if ((alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE) == 0) {
+		ticks += current;
+	}
+
+	/* this timer always counts up. */
+	config->base->CHANNEL[config->channel].COMP1 = ticks;
+
+	data->interrupt_mask |= kQTMR_Compare1InterruptEnable;
+	QTMR_EnableInterrupts(config->base, config->channel, data->interrupt_mask);
+
+	return 0;
+}
+
+static int mcux_qtmr_cancel_alarm(const struct device *dev, uint8_t chan_id)
+{
+	const struct mcux_qtmr_config *config = dev->config;
+	struct mcux_qtmr_data *data = dev->data;
+
+	if (chan_id != 0) {
+		LOG_ERR("Invalid channel id");
+		return -EINVAL;
+	}
+
+	QTMR_DisableInterrupts(config->base, config->channel, data->interrupt_mask);
+	data->interrupt_mask &= ~kQTMR_Compare1InterruptEnable;
+	data->alarm_callback = NULL;
+
+	return 0;
+}
+
+static uint32_t mcux_qtmr_get_pending_int(const struct device *dev)
+{
+	const struct mcux_qtmr_config *config = dev->config;
+
+	return QTMR_GetStatus(config->base, config->channel);
+}
+
+static int mcux_qtmr_set_top_value(const struct device *dev,
+				  const struct counter_top_cfg *cfg)
+{
+	const struct mcux_qtmr_config *config = dev->config;
+	struct mcux_qtmr_data *data = dev->data;
+
+	if (cfg->ticks != config->info.max_top_value) {
+		LOG_ERR("Wrap can only be set to 0x%x",
+			config->info.max_top_value);
+		return -ENOTSUP;
+	}
+
+	if ((cfg->flags & COUNTER_TOP_CFG_DONT_RESET) == 0) {
+		if ((config->base->CHANNEL[config->channel].CTRL & TMR_CTRL_DIR_MASK) != 0U) {
+			/* counting down, reset to UINT16MAX */
+			config->base->CHANNEL[config->channel].CNTR = UINT16_MAX;
+		} else {
+			/* counting up, reset to 0 */
+			config->base->CHANNEL[config->channel].CNTR = 0;
+		}
+	}
+
+	if (cfg->callback != NULL) {
+		data->top_callback = cfg->callback;
+		data->top_user_data = cfg->user_data;
+
+		data->interrupt_mask |= kQTMR_OverflowInterruptEnable;
+		QTMR_EnableInterrupts(config->base, config->channel, kQTMR_OverflowInterruptEnable);
+	}
+
+	return 0;
+}
+
+static uint32_t mcux_qtmr_get_top_value(const struct device *dev)
+{
+	const struct mcux_qtmr_config *config = dev->config;
+
+	return config->info.max_top_value;
```

Comment:
If `set_top_value` is set, then this should return that as the `top_value`

Reply:
The QTMR does not support setting a top value other than the one stored in `config->info.max_top_value` (which is 0xffff).

The implementation of `set_top_value` for this driver returns `-ENOTSUP` for any other value. As far as I understand, that's the default behaviour in this case.

Reply:
Given there are 4 channels, could we reserve one of the channels for `set_top_value` implementation and take out one from the device tree file.